### PR TITLE
Add batch norm for 2d and 4d inputs + tests

### PR DESCRIPTION
**Pull request by tomlepaine**
Added a batch norm layer helper that works for both 2d and 4d input, in the style of #7.

Using it would look something like:
```
# 2d example
x = tf.placeholder(tf.float32, [batch, dim])
b = batch_norm(x, 'b')

# 4d example
x = tf.placeholder(tf.float32, [batch, width, height, dim])
b = batch_norm(x, 'b')
```

---

I also wrote a simple test that I guess does two things:
- checks that the layers don't throw any exceptions.
- checks the shape out the output.

### CODE REVIEW

**Comment by kevjshih** [comment]
Should we optionally give it saved mean and variances as inputs for test-time use? Also we should probably output the means and variances so we can save them for test time.

File: aiutils/tftools/layers.py [hunk 1]
```diff
@@ -81,3 +81,41 @@ def conv2d(input,
...
+        Will have the same shape as input.
+    """
+    rank = len(input.get_shape().as_list())
+    in_dim = input.get_shape().as_list()[-1]
+    if rank == 2:
+        axes = [0]
+    elif rank == 4:
+        axes = [0, 1, 2]
+    else:
+        raise ValueError('rank must be 2 or 4.')
```
Comment (BigRedT) [inline aiutils/tftools/layers.py]: Probably assert is more appropriate here than raising exception. You can use the devtool for this.

**Comment by tomlepaine** [comment]
@kevjshih How do you normally use batch norm at test time? I am usually lazy and just run it on batches without any modification. This gives reasonable results, but I guess they aren't optimal.

I'm told the _correct_ way is to calculate the mean and variance on the whole training set and use that at test time, or something like that.

If you think outputting the mean and variance is the best way to use this at test time, that works for me.

File: aiutils/tftools/layers.py (hunk 1 above)
Comment (tomlepaine) [inline aiutils/tftools/layers.py]: I was thinking that too, but the `devtools` weren't checked in yet 😎.  I can switch over now.

**Comment by kevjshih** [comment]
Yeah, so I've done that before because that was the old matconvnet implementation (currently fixed), but technically it counts as unsupervised learning on the test set, which may or may not be bad depending on your point of view. Also it would cause things to break if your batch size is too small, causing the estimated mean and variance to be too different from what it had trained on. It'll definitely fail for batch sizes of 1.

File: aiutils/tftools/layers.py (hunk 1 above)
Comment (BigRedT) [inline aiutils/tftools/layers.py]: cool.

**Comment by BigRedT** [comment]
Another way to use batch norm at test is to keep an exponential running average using training and use that during test. Jason has some experience with that.

**Comment by jrock08** [comment]
I'll put something up later that I use.  This is pretty reasonable as a "good enough for government work" thing

File: aiutils/tftools/layers.py (hunk 1 above)
Comment (kevjshih) [inline aiutils/tftools/layers.py]: I think an assertion before the conditional is equivalent to the current code, since all assertions do is to raise a different type of exception in the end. It would add a redundant check for 2 or 4 dimensions instead. 

On the other hand, if you want to throw it into the logs, just add an `import logging` at the top of the file and maybe add a log entry here before 109 with something like `logging.error("rank was not 2 or 4")`

Comment (BigRedT) [inline aiutils/tftools/layers.py]: i mean't remove the exception and add assert in the beginning. What is the purpose of putting it in tensorflow logs?

Comment (kevjshih) [inline aiutils/tftools/layers.py]: Well, if you put the assertion at the beginning, you'll essentially have:
```python
assert rank == 2 or rank ==4
if rank == 2:
        axes = [0]
elif rank == 4:
        axes = [0, 1, 2]
```

So if the assertion is passed, you end up checking the same thing twice. In the current implementation, you only check once.

Comment (BigRedT) [inline aiutils/tftools/layers.py]: I think the idea is to make clear what your assumptions about inputs are as soon as possible, rather than have a user read through the logic and then understand it indirectly. When computation isn't that expensive, readability should be given preference.  Though here it doesn't seem to be that big a deal either way.

Comment (kevjshih) [inline aiutils/tftools/layers.py]: True, but in that sense, reading line 109 should be pretty clear already. I don't think moving it up will make a difference.
